### PR TITLE
[Enhancement] Support struct cast function

### PR DESCRIPTION
**Pull request by Seaven**
Fixes #issue

* check and update struct cast function
* support index expression for struct
* add struct test

```

MySQL tpcds_1g> select row(1, 2, 3, 4)[1]
+--------------------+
| row(1, 2, 3, 4)[1] |
+--------------------+
|         1          |
+--------------------+

MySQL tpcds_1g> select cast(row(1, 2) as STRUCT<a string, b string>);
+-----------------------------------------------------------------+
| CAST((row(1, 2)) AS struct<a varchar(65533), b varchar(65533)>) |
+-----------------------------------------------------------------+
| {"a":"1","b":"2"}                                               |
+-----------------------------------------------------------------+
1 row in set
```

## What type of PR is this:
- [ ] BugFix
- [ ] Feature
- [x] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

## Checklist:
- [x] I have added test cases for my bug fix or my new feature
- [ ] This pr will affect users' behaviors
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function

## Bugfix cherry-pick branch check:
- [x] I have checked the version labels which the pr will be auto-backported to the target branch
  - [ ] 3.0
  - [ ] 2.5
  - [ ] 2.4
  - [ ] 2.3

### CODE REVIEW

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 63 / 69 (91.30%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/optimizer/rewrite/scalar/NormalizePredicateRule.java|15|18|83.33%|[209, 210, 218]|
|:large_blue_circle:|com/starrocks/sql/analyzer/QueryAnalyzer.java|20|22|90.91%|[637, 645]|
|:large_blue_circle:|com/starrocks/sql/analyzer/ExpressionAnalyzer.java|17|18|94.44%|[465]|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/scalar/SubfieldOperator.java|6|6|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/common/TypeManager.java|3|3|100.00%|[]|
|:large_blue_circle:|com/starrocks/catalog/StructType.java|2|2|100.00%|[]|

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rewrite/scalar/NormalizePredicateRule.java [hunk 1]
```diff
@@ -187,4 +193,35 @@ public ScalarOperator visitInPredicate(InPredicateOperator predicate, ScalarOper
...
         return isIn ? Utils.compoundOr(result) : Utils.compoundAnd(result);
     }
+
+    // rewrite collection element to subfiled
+    @Override
+    public ScalarOperator visitCollectionElement(CollectionElementOperator collectionElement,
+                                                 ScalarOperatorRewriteContext context) {
+        if (collectionElement.getChild(0).getType().isStructType()) {
+            Preconditions.checkState(collectionElement.getChild(1).isConstantRef());
+            Preconditions.checkState(collectionElement.getChild(1).getType().isIntegerType());
```
Comment (packy92) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rewrite/scalar/NormalizePredicateRule.java:203]: Is these check necessary?

File: be/src/exprs/cast_expr.cpp [hunk 2]
```diff
@@ -1621,6 +1621,9 @@ StatusOr<std::unique_ptr<Expr>> VectorizedCastExprFactory::create_cast_expr(Obje
         if (from_type.children.size() != to_type.children.size()) {
             return Status::NotSupported("vectorized engine not support cast struct with different number of children.");
         }
+        if (to_type.field_names.empty() || from_type.field_names.size() != to_type.field_names.size()) {
```
Comment (packy92) [inline be/src/exprs/cast_expr.cpp:1624]: FE has similar check. So is it for stablity?

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rewrite/scalar/NormalizePredicateRule.java (hunk 1 above)
Comment (Seaven) [reply fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rewrite/scalar/NormalizePredicateRule.java:203]: emmm, I think unnecessary, just avoid person write error code

File: be/src/exprs/cast_expr.cpp (hunk 2 above)
Comment (Seaven) [reply be/src/exprs/cast_expr.cpp:1624]: check strict, avoid crash in upgrade

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 63 / 69 (91.30%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/optimizer/rewrite/scalar/NormalizePredicateRule.java|15|18|83.33%|[209, 210, 218]|
|:large_blue_circle:|com/starrocks/sql/analyzer/QueryAnalyzer.java|20|22|90.91%|[637, 645]|
|:large_blue_circle:|com/starrocks/sql/analyzer/ExpressionAnalyzer.java|17|18|94.44%|[465]|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/scalar/SubfieldOperator.java|6|6|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/common/TypeManager.java|3|3|100.00%|[]|
|:large_blue_circle:|com/starrocks/catalog/StructType.java|2|2|100.00%|[]|

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rewrite/scalar/NormalizePredicateRule.java (hunk 1 above)
Comment (packy92) [reply fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rewrite/scalar/NormalizePredicateRule.java:203]: Maybe add some error msg is better? If not, the unknown error return to client is really confused.

**Review by packy92 — APPROVED** [review]

**Review by liuyehcf — APPROVED** [review]

**Pull request merged**